mod/consensus/pkg/cometbft/service/abci.go

```diff
@@ -90,75 +89,76 @@ func (s *Service[LoggerT]) InitChain(
 		// InitChain represents the state of the application BEFORE the first
 		// block, i.e. the genesis block. This means that when processing the
 		// app's InitChain handler, the block height is zero by default.
-		// However, after Commit is called
-		// the height needs to reflect the true block height.
-		initHeader.Height = req.InitialHeight
+		// However, after genesis is handled the height needs to reflect
+		// the true block height.
+		initHeader := cmtproto.Header{
+			ChainID: req.ChainId,
+			Time:    req.Time,
+			Height:  req.InitialHeight,
+		}
 		s.finalizeBlockState.SetContext(
 			s.finalizeBlockState.Context().WithBlockHeader(initHeader),
 		)
 	}()
 
-	if s.finalizeBlockState == nil {
-		return nil, errors.New("finalizeBlockState is nil")
-	}
-
-	res, err := s.initChainer(s.finalizeBlockState.Context(), req)
+	resValidators, err := s.initChainer(
+		s.finalizeBlockState.Context(),
+		req.AppStateBytes,
+	)
 	if err != nil {
 		return nil, err
 	}
 
-	if res == nil {
+	if len(resValidators) == 0 {
 		return nil, errors.New(
-			"application init chain handler returned a nil response",
+			"application init chain handler returned a no validators",
 		)
 	}
 
 	if len(req.Validators) > 0 {
-		if len(req.Validators) != len(res.Validators) {
+		if len(req.Validators) != len(resValidators) {
 			return nil, fmt.Errorf(
 				"len(RequestInitChain.Validators) != len(GenesisValidators) (%d != %d)",
 				len(req.Validators),
-				len(res.Validators),
+				len(resValidators),
 			)
 		}
 
 		sort.Sort(cmtabci.ValidatorUpdates(req.Validators))
 
-		for i := range res.Validators {
-			if req.Validators[i].Power != res.Validators[i].Power {
+		for i := range resValidators {
+			if req.Validators[i].Power != resValidators[i].Power {
 				return nil, errors.New("mismatched power")
 			}
 			if !bytes.Equal(
-				req.Validators[i].PubKeyBytes, res.Validators[i].
+				req.Validators[i].PubKeyBytes, resValidators[i].
 					PubKeyBytes) {
 				return nil, errors.New("mismatched pubkey bytes")
 			}
 
-			if req.
-				Validators[i].PubKeyType != res.
-				Validators[i].PubKeyType {
+			if req.Validators[i].PubKeyType !=
+				resValidators[i].PubKeyType {
 				return nil, errors.New("mismatched pubkey types")
 			}
 		}
 	}
 
 	// NOTE: We don't commit, but FinalizeBlock for block InitialHeight starts
-	// from
-	// this FinalizeBlockState.
+	// from this FinalizeBlockState.
 	return &cmtabci.InitChainResponse{
-		ConsensusParams: res.ConsensusParams,
```

Comment:
`ConsensusParams` is not set by the response generated in `initChainer`,  so this effectively assigning default value of ConsensusParams.
Replaced with ConsensusParams from request, which I believe is what we need to do